src/text.rs

```diff
@@ -0,0 +1,210 @@
+    /// Renders multiple lines from left-to-right according to the inner alignment.
+    Wrap(Align),
+}
+
+impl Default for Layout {
+    fn default() -> Self {
+        Layout::SingleLine(Align::Left)
+    }
+}
+
+impl From<Align> for gfx_glyph::HorizontalAlign {
+    fn from(align: Align) -> gfx_glyph::HorizontalAlign {
+        match align {
+            Align::Left => gfx_glyph::HorizontalAlign::Left,
+            Align::Center => gfx_glyph::HorizontalAlign::Center,
+            Align::Right => gfx_glyph::HorizontalAlign::Right,
+        }
+    }
+}
+
+impl From<Layout> for gfx_glyph::Layout<StandardLineBreaker> {
+    fn from(layout: Layout) -> gfx_glyph::Layout<StandardLineBreaker> {
+        match layout {
+            Layout::Wrap(a) => gfx_glyph::Layout::Wrap(StandardLineBreaker, a.into()),
+            Layout::SingleLine(a) =>
+                gfx_glyph::Layout::SingleLine(StandardLineBreaker, a.into()),
+        }
+    }
+}
+
+/// Smart pointer containing a font to draw text.
+#[derive(Clone)]
+pub struct Font{
+    brush: Rc<RefCell<GlyphBrush<'static, BackendResources, BackendFactory>>>,
+    pub(crate) path: PathBuf,
+}
+
+impl Font {
+    pub(crate) fn new(buf: Vec<u8>, path: PathBuf, factory: BackendFactory) -> Font {
+        Font {
+            brush: Rc::new(RefCell::new(GlyphBrushBuilder::using_font(buf).build(factory))),
+            path: path,
+        }
+    }
+
+    pub(crate) fn queue(&self, section: &OwnedSection, layout: Layout) {
+        let mut brush = self.brush.borrow_mut();
+        let layout: gfx_glyph::Layout<StandardLineBreaker> = layout.into();
+        brush.queue(section, &layout);
+    }
+
+    pub(crate) fn draw(&self,
+                       encoder: &mut Encoder<BackendResources, BackendCommandBuffer>,
+                       out: &RenderTargetView<BackendResources, ColorFormat>) {
+        let mut brush = self.brush.borrow_mut();
+        brush.draw_queued(encoder, out).expect("Error while drawing text");
+    }
+}
+
+impl fmt::Debug for Font {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Font {{ path: {:?} }}", self.path)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TextData {
+    pub(crate) section: OwnedSection,
+    pub(crate) layout: Layout,
+    pub(crate) font: Font,
+}
+
+impl TextData {
+    pub(crate) fn new<S:Into<String>>(font: &Font, text: S) -> Self {
+        TextData {
+            section: OwnedSection {
+                text: text.into(),
+                color: [1.0, 1.0, 1.0, 1.0],
+                .. OwnedSection::default()
+            },
+            layout: Layout::default(),
+            font: font.clone(),
+        }
+    }
+}
+
+/// UI (on-screen) text.
+/// To use, create the new one using [`Factory::ui_text`](struct.Factory.html#method.ui_text)
+/// and add it to the scene using [`Scene::add`](struct.Scene.html#method.add).
+#[derive(Debug, Clone)]
+pub struct Text {
+    pub(crate) object: Object,
+    pub(crate) data: TextData,
```

Comment:
note: carrying the copy of internal data with the local object is somewhat an anti-pattern, as you can see nothing else doing it, because otherwise animating those properties would be hard (w.r.t. synchronization between this local state and the remote state in the hub)

Reply:
Ugh, I know. I can change it, but it will lead us to massive bunch of pattern matching in Hub.

Reply:
Nevermind, I will fix that.